PXC-3691: garbd doesn't correctly handle multiple incoming SST requests at the same time

voodam opened the pull request:
Related PR: https://github.com/percona/percona-xtradb-cluster/pull/1602
This is backporting issue. C++11 requirement wasn't added, pthreads primitives used instead just like in 3-x branch.

voodam committed:
PXC-3691: garbd doesn't correctly handle multiple incoming SST requests at the same time

Backporting PXC-2604 to 5.7

voodam committed:
PXC-3691: garbd doesn't correctly handle multiple incoming SST requests at the same time

Backporting PXC-3031 to 5.7.
Also dirty fixed gcs_get_state_for_idx: for unexisting node it returns GCS_NODE_STATE_MAX instead of crashing from segfault.

voodam committed:
PXC-3691: garbd doesn't correctly handle multiple incoming SST requests at the same time

Backporting PXC-3353 to 5.7

voodam committed:
PXC-3691: garbd doesn't correctly handle multiple incoming SST requests at the same time

Backporting PXC-3538 to 5.7

voodam committed:
PXC-3691: garbd doesn't correctly handle multiple incoming SST requests at the same time

Backporting PXC-3416 to 5.7

voodam committed:
PXC-3691: garbd doesn't correctly handle multiple incoming SST requests at the same time

Backporting PXC-3404 to 5.7

dutow commented on gcs/src/gcs.cpp:
```diff
@@ -199,6 +199,20 @@ struct gcs_conn
+gcs_node_state_t gcs_get_state_for_idx(gcs_conn_t* conn, ssize_t idx) {
+  const gcs_group_t* group = gcs_core_get_group(conn->core);
+  if (idx >= group->num) {
+    // XXX: Dirty fix.
```
This looks like a real issue that could also happen for 8.0, we just didn't notice it yet - and we do need this condition here as a safeguard.

On the other hand, it is not the full fix for this, I'll create a separate jira issue about that for us, and this condition is good enough for now, you can remove the comment from it.

voodam replied:
Ok, but in fairness, for some reason this segfault is not reproduced in 8.0.
Maybe it would be better to use optional (like boost optional or something)? Or introduce new GCS_NODE_STATE_* status. GCS_NODE_STATE_MAX seems not to be a good choice here.

dutow replied:
The main issue here is that we keep using the `idx` we obtained at the start of the SST request, and there's no guarantee that the node array doesn't change during it. That's also true for 8.0.

For now this fix, an optional could work, yes, and would be less of a hack.

voodam replied:
Done

dutow commented on gcs/src/gcs_core.cpp:
```diff
@@ -999,12 +999,12 @@ core_msg_to_action (gcs_core_t*          core,
-                // Also note that original ret value is not preserved on return
-                // so this must be done here.
-                gu_abort();
+                // A negative return value ends the receive loop, as this is
```
This should be an `ifdef GCS_FOR_GARB` change, as in 8.0.

voodam replied:
Done

voodam committed:
core_msg_to_action fix

voodam committed:
PXC-3691: garbd doesn't correctly handle multiple incoming SST requests at the same time

Backporting PXC-3544 to 5.7

voodam commented:
Some minor fixes added.

voodam committed:
PXC-3691: garbd doesn't correctly handle multiple incoming SST requests at the same time

Backporting PXC-3909 to 5.7

voodam committed:
PXC-3691: garbd doesn't correctly handle multiple incoming SST requests at the same time

Backporting PXC-3909 to 5.7
Fix: use the correct return value even in the terminate case

kamil-holubicki commented on galera/src/replicator_smm_stats.cpp:
```diff
@@ -256,7 +256,7 @@ galera::ReplicatorSMM::stats_get()
     std::ostringstream osinterval;
     osinterval << "[ " << stats.fc_lower_limit << ", " << stats.fc_upper_limit << " ]";
-    strncpy(interval_string_, osinterval.str().c_str(), sizeof(interval_string_));
+    strncpy(interval_string_, osinterval.str().c_str(), sizeof(interval_string_)-1);
```
I think it should be followed by
`interval_string_[sizeof(interval_string_)-1] = 0;`

kamil-holubicki commented on galera/src/replicator_smm_stats.cpp:
```diff
@@ -319,7 +319,7 @@ galera::ReplicatorSMM::stats_get()
         os << percent_complete << "% complete, received seqno "
            << current << " of " << first << "-" << last;
-        strncpy(ist_status_string_, os.str().c_str(), sizeof(ist_status_string_));
+        strncpy(ist_status_string_, os.str().c_str(), sizeof(ist_status_string_)-1);
```
likewise

kamil-holubicki reviewed (approved):
lgtm
